src/pudl/metadata/classes.py

```diff
@@ -1748,3 +1749,58 @@ def to_rst(self, top_dir: DirectoryPath, csv_subdir: DirectoryPath, rst_path: st
                 rendered = encoder.to_rst(
                     top_dir=top_dir, csv_subdir=csv_subdir, is_header=header)
                 f.write(rendered)
+
+
+class DatasetteMetadata(Base):
+    """
+    A collection of Data Sources and Resources for metadata export.
+
+    Used to create metadata YAML file to accompany Datasette.
+    """
+
+    datasource_dict: Dict[str, DataSource]
+    resource_package: Package = Package.from_resource_ids()
+    label_column_dict: Dict[str, str]
+    years_dict: Dict[str, Any]
+
+    @classmethod
+    def from_data_sources(
+        cls, data_sources: Iterable[str],
+        label_column_dict: Dict[str, str]
+    ) -> "DatasetteMetadata":
+        """
+        Construct a dictionary of DataSources from data source names.
+
+        Create dictionary of first and last year or year-month for each source.
+
+        Args:
+            data_sources: data sources currently included in Datasette
+            label_column_dict: map from resource to column name for link label
+        """
+        datasource_dict = {}
+        years_dict = {}
+        for ds_id in data_sources:
+            source = DataSource.from_id(ds_id)
+            datasource_dict[ds_id] = source
+            if 'years' in source.working_partitions.keys():
+                last_idx = len(source.working_partitions['years']) - 1
+                years_dict[ds_id] = [
+                    source.working_partitions['years'][0],
+                    source.working_partitions['years'][last_idx]]
```

Comment:
I think there's a better way to handle displaying the "temporal coverage" of the different data sources.
* We already know that we're going to want to be able to display this information in other contexts (e.g. for the data source documentation).
* We also know that the years in `working_partitions` isn't quite the same thing as the same thing as the time span covered by the data source. The working partitions are really about how the data gets chunked up for processing. For example the FERC 714 is a single monolithic input, but it still covers the time from 2006-2020.
* Together, this suggests to me that the right place to put this display logic is probably the `DataSource` class itself, so it can be re-used whenever we want to display information about what time a data source covers, and not have to deal with it multiple times in different output contexts.

So you might define a `DataSource.temporal_coverage()` method that potentially has different implementations for the different data sources. For an annually partitioned data source like FERC 1, it might return the string "1994-2020" which could be as simple as:
```py
return f"{min(working_partitions['years']}-{max(working_partitions['years'])}"
```
For the FERC 714 (eventually) we would need to have some other internal data that's not linked to the working partitions, but the externally exposed method would be the same, and implement the same idea. For the a dataset like the `eia860m` which only ever gives a single snapshot in time, it might return something like "through 2021-08"

I think arranging it like that would reduce the complexity of this code and keep the irreducible complexity close to where it really lives -- inside the data source itself, which should know how to deal with its own mess internally.

Separately, note you can index lists from either the beginning or the end:
```py
first_year = years[0]
last_year = years[-1]  # negative indices count from the end of the list.
```
In this particular case (because of the validation on the `DataSource` we know that the list of years is sorted, but if we didn't know that you could also do
```py
first_year = min(years)
last_year = max(years)
```